In [142]:
#Comenzaremos importando las librerias necesarias para manipular la informacion

import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns

In [143]:
#Importamos los datasets provistos para el MVP

data_movies = pd.read_csv('movies_dataset.csv')

data_credits = pd.read_csv('credits.csv')

C:\Users\Alejandro\AppData\Local\Temp\ipykernel_14068\2846471490.py:3: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  data_movies = pd.read_csv('movies_dataset.csv')


In [144]:
#Damos un vistazo a los datasets
data_movies.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2  A family wedding reignites the ancient feud be...  ...   1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4  Just when George Banks has recovered from his ...  ...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

In [145]:
data_credits.head()

cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   

                                                crew     id  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...    862  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   8844  
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...  15602  
3  [{'credit_id': '52fe44779251416c91011acb', 'de...  31357  
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...  11862

   ETL (Extraccion, transformacion y limpieza)

 PRIMERAS TRANSFORMACIONES

Comenzaremos realizando las primeras transformaciones exigidas en las cosignas de este MVP 

In [146]:
'''Se comenzará con las consignas indicadas para realizar el MVP, a continuación se rellenaran los valores nulos de 'reveneu' y 'budget' con 0
  luego se convertiran todos sus valores a float para luego poder crear la columna return'''
#Primero debemos asegurarnos que todos los valores sean numericos en ambas columnas, si hay valores que no se puedan convertir quedaran como NaN
data_movies['revenue'] = pd.to_numeric(data_movies['revenue'], errors='coerce')
data_movies['budget'] = pd.to_numeric(data_movies['budget'], errors='coerce')
#Luego se rellenaran los valores nulos con 0, y aplicamos la conversion a float
data_movies['revenue'] = round(data_movies['revenue'].fillna(0).astype(float),2)
data_movies['budget'] = round(data_movies['budget'].fillna(0).astype(float),2)

In [147]:
#Creamos la columna "return" asegurandonos que en los casos donde budget se 0 solo devuelva el valor 0 y no un NaN
data_movies['return'] = round(data_movies.apply(lambda row: row['revenue'] / row['budget'] if row['budget'] != 0 else 0, axis=1),2)

In [148]:
'''Pasamos a la consigna donde tratamos la columna "release_date", le daremos formato AA-mm-dd con "to_datetime" y, con el parametro
errors='coerce' convertiremos los valores no validos a NaT. Luego eliminaremos todos los valores nulos y, posteriormente
crearemos la columna "release_year"'''
#Damos el formato AA-mm-dd
data_movies['release_date'] = pd.to_datetime(data_movies['release_date'], format='%Y-%m-%d', errors='coerce')
#Eliminamos los valores nulos
data_movies.dropna(subset=['release_date'], inplace=True)
#Creamos la columna "release_year"
data_movies['release_year'] = data_movies['release_date'].dt.year

In [149]:
#Eliminamos las columnas que explicitaron en las consignas que no serviran
data_movies = data_movies.drop(columns=['adult','homepage','imdb_id','original_title','poster_path','video'])

Continuaremos con el ETL pero mientras realizaremos un EDA ligero para detectar posibles transformaciones rápidas de modo intuitivo, por ejempo revision a simple vista de nulos y otros.

In [150]:
data_movies.info()

<class 'pandas.core.frame.DataFrame'>
Index: 45376 entries, 0 to 45465
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   belongs_to_collection  4488 non-null   object        
 1   budget                 45376 non-null  float64       
 2   genres                 45376 non-null  object        
 3   id                     45376 non-null  object        
 4   original_language      45365 non-null  object        
 5   overview               44435 non-null  object        
 6   popularity             45376 non-null  object        
 7   production_companies   45376 non-null  object        
 8   production_countries   45376 non-null  object        
 9   release_date           45376 non-null  datetime64[ns]
 10  revenue                45376 non-null  float64       
 11  runtime                45130 non-null  float64       
 12  spoken_languages       45376 non-null  object        
 13  status

- Podemos notar que las columnas "belongs_to_collection" tiene casi el 100% de nulos entonces procederemos a eliminarla
- De misma forma observamos que columna "tagline" tiene aproximadamente el 50% de sus valores nulos y, de acuerdo a lo que vimos en descripcion de datos es solo una frase celebre de la pelicula por lo tanto, procederemos a eliminarla porque no proporciona informacion relevane para este MVP y posee muchos nulos.
- Se observa que "id" y "popularity" estan en formato texto y son caracteristivas numericas, por lo tanto hay que modificarlos.

In [151]:
#Eliminamos "tagline" y "belongs_to_conection"
data_movies = data_movies.drop(columns=['belongs_to_collection', 'tagline'])

#Cambiamos formato a "id" y "popularity"
data_movies['id'] = data_movies['id'].astype(int)
data_movies['popularity'] = round(data_movies['popularity'].astype(float),2)

In [152]:
#Comenzaremos con las columnas anidadas, en este caso la columna "genres", ahora exploramos los datos
data_movies['genres'].head()

0    [{'id': 16, 'name': 'Animation'}, {'id': 35, '...
1    [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...
2    [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...
3    [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...
4                       [{'id': 35, 'name': 'Comedy'}]
Name: genres, dtype: object

In [153]:
#Podemos ver que son diccionarios de generos anidados dentro de estructura de listas. Los diccionarios solo tienen el "id" y "name"
#de los géneros, solo vamos a conservar el nombre de estos ya que el id no nos proporciona informacion
data_movies['genres'] = data_movies['genres'].fillna('[]').apply(eval)
#Hacemos una funcion que nos permita extraer el nombre de los géneros
def extraer_nombres(column):
    return [item['name'] for item in column]
#Creamos una columna "name_genre" donde se guardarán los nombres de las géneros asociados
data_movies['genres'] = data_movies['genres'].apply(extraer_nombres)
#Reemplazamos las listas vacias nuevamente por NaN
data_movies['genres'] = data_movies['genres'].replace([], np.nan)#Esto nos servira para indentificar datos nulos o vacios en el EDA

In [154]:
#Ahora vamos a continuar por la columna "production_companies", la visualizamos y exploramos, podemos notar que su composición es similar a
#la columna "genres", por lo tanto vamos a seguir los mismos pasos y solo quedarnos con el nombre de las empresas productoras.
data_movies['production_companies'].head()

0       [{'name': 'Pixar Animation Studios', 'id': 3}]
1    [{'name': 'TriStar Pictures', 'id': 559}, {'na...
2    [{'name': 'Warner Bros.', 'id': 6194}, {'name'...
3    [{'name': 'Twentieth Century Fox Film Corporat...
4    [{'name': 'Sandollar Productions', 'id': 5842}...
Name: production_companies, dtype: object

In [155]:
#Cambiamos nulos por listas vacias y evaluamos como diccionarios
data_movies['production_companies'] = data_movies['production_companies'].fillna('[]').apply(eval)
#Creamos una columna "name_companie" donde se guardarán los nombres de las empresas productoras, aplicamos la función para extraer nombres
data_movies['production_companies'] = data_movies['production_companies'].apply(extraer_nombres)
#Reemplazamos las listas vacias nuevamente por NaN
data_movies['production_companies'] = data_movies['production_companies'].replace([], np.nan)

In [156]:
#Ahora vamos a continuar por la columna "production_countries", la visualizamos y exploramos, podemos notar que su composición es similar a
#la columna "genres", por lo tanto vamos a seguir los mismos pasos y solo quedarnos con el nombre de los paises.
data_movies['production_countries'].head()

0    [{'iso_3166_1': 'US', 'name': 'United States o...
1    [{'iso_3166_1': 'US', 'name': 'United States o...
2    [{'iso_3166_1': 'US', 'name': 'United States o...
3    [{'iso_3166_1': 'US', 'name': 'United States o...
4    [{'iso_3166_1': 'US', 'name': 'United States o...
Name: production_countries, dtype: object

In [157]:
#Cambiamos nulos por listas vacias y evaluamos como diccionarios
data_movies['production_countries'] = data_movies['production_countries'].fillna('[]').apply(eval)
#Creamos una columna "name_companie" donde se guardarán los nombres de las empresas productoras, aplicamos la función para extraer nombres
data_movies['production_countries'] = data_movies['production_countries'].apply(extraer_nombres)
#Reemplazamos las listas vacias nuevamente por NaN
data_movies['production_countries'] = data_movies['production_countries'].replace([], np.nan)

In [158]:
#Seguimos con la columna "spoken_languages", la visualizamos y exploramos, podemos notar que su composición es similar a
#la columna "genres", por lo tanto vamos a seguir los mismos pasos y solo quedarnos con el nombre de los idiomas.
data_movies['spoken_languages'].head()

0             [{'iso_639_1': 'en', 'name': 'English'}]
1    [{'iso_639_1': 'en', 'name': 'English'}, {'iso...
2             [{'iso_639_1': 'en', 'name': 'English'}]
3             [{'iso_639_1': 'en', 'name': 'English'}]
4             [{'iso_639_1': 'en', 'name': 'English'}]
Name: spoken_languages, dtype: object

In [159]:
#Cambiamos nulos por listas vacias y evaluamos como diccionarios
data_movies['spoken_languages'] = data_movies['spoken_languages'].fillna('[]').apply(eval)
#Creamos una columna "name_companie" donde se guardarán los nombres de las empresas productoras, aplicamos la función para extraer nombres
data_movies['spoken_languages'] = data_movies['spoken_languages'].apply(extraer_nombres)
#Reemplazamos las listas vacias nuevamente por NaN
data_movies['spoken_languages'] = data_movies['spoken_languages'].replace([], np.nan)

In [160]:
#Visualizamos los datos resultantes de las modificaciones en data_movies
data_movies.head(10)

budget                              genres     id original_language  \
0  30000000.0         [Animation, Comedy, Family]    862                en   
1  65000000.0        [Adventure, Fantasy, Family]   8844                en   
2         0.0                   [Romance, Comedy]  15602                en   
3  16000000.0            [Comedy, Drama, Romance]  31357                en   
4         0.0                            [Comedy]  11862                en   
5  60000000.0    [Action, Crime, Drama, Thriller]    949                en   
6  58000000.0                   [Comedy, Romance]  11860                en   
7         0.0  [Action, Adventure, Drama, Family]  45325                en   
8  35000000.0       [Action, Adventure, Thriller]   9091                en   
9  58000000.0       [Adventure, Action, Thriller]    710                en   

                                            overview  popularity  \
0  Led by Woody, Andy's toys live happily in his ...       21.95   
1  When siblings Judy and Peter discover an encha...       17.02   
2  A family wedding reignites the ancient feud be...       11.71   
3  Cheated on, mistreated and stepped on, the wom...        3.86   
4  Just when George Banks has recovered from his ...        8.39   
5  Obsessive master thief, Neil McCauley leads a ...       17.92   
6  An ugly duckling having undergone a remarkable...        6.68   
7  A mischievous young boy, Tom Sawyer, witnesses...        2.56   
8  International action superstar Jean Claude Van...        5.23   
9  James Bond must unmask the mysterious head of ...       14.69   

                                production_companies  \
0                          [Pixar Animation Studios]   
1  [TriStar Pictures, Teitler Film, Interscope Co...   
2                     [Warner Bros., Lancaster Gate]   
3           [Twentieth Century Fox Film Corporation]   
4       [Sandollar Productions, Touchstone Pictures]   
5  [Regency Enterprises, Forward Pass, Warner Bros.]   
6  [Paramount Pictures, Scott Rudin Productions, ...   
7                             [Walt Disney Pictures]   
8  [Universal Pictures, Imperial Entertainment, S...   
9                  [United Artists, Eon Productions]   

                         production_countries release_date      revenue  \
0                  [United States of America]   1995-10-30  373554033.0   
1                  [United States of America]   1995-12-15  262797249.0   
2                  [United States of America]   1995-12-22          0.0   
3                  [United States of America]   1995-12-22   81452156.0   
4                  [United States of America]   1995-02-10   76578911.0   
5                  [United States of America]   1995-12-15  187436818.0   
6         [Germany, United States of America]   1995-12-15          0.0   
7                  [United States of America]   1995-12-22          0.0   
8                  [United States of America]   1995-12-22   64350171.0   
9  [United Kingdom, United States of America]   1995-11-16  352194034.0   

   runtime             spoken_languages    status  \
0     81.0                    [English]  Released   
1    104.0          [English, Français]  Released   
2    101.0                    [English]  Released   
3    127.0                    [English]  Released   
4    106.0                    [English]  Released   
5    170.0           [English, Español]  Released   
6    127.0          [Français, English]  Released   
7     97.0           [English, Deutsch]  Released   
8    106.0                    [English]  Released   
9    130.0  [English, Pусский, Español]  Released   

                         title  vote_average  vote_count  return  release_year  
0                    Toy Story           7.7      5415.0   12.45          1995  
1                      Jumanji           6.9      2413.0    4.04          1995  
2             Grumpier Old Men           6.5        92.0    0.00          1995  
3            Waiting to Exhale           6.1    

In [161]:
#Ahora pasaremos al df data_credits cuyas columnas "cast" y "crew" también tienen datos anidados.
#Cambiamos nulos por listas vacias y evaluamos como diccionarios los datos de "cast"
data_credits['cast'] = data_credits['cast'].fillna('[not found]').apply(eval)

#Ahora crearemos un nuevo df para desanidar la columna "cast" 
df_cast = pd.json_normalize(data_credits['cast'].explode())

In [162]:
#Visualizamos el df de los actores
df_cast

cast_id                character                 credit_id  gender  \
0          14.0            Woody (voice)  52fe4284c3a36847f8024f95     2.0   
1          15.0   Buzz Lightyear (voice)  52fe4284c3a36847f8024f99     2.0   
2          16.0  Mr. Potato Head (voice)  52fe4284c3a36847f8024f9d     2.0   
3          17.0       Slinky Dog (voice)  52fe4284c3a36847f8024fa1     2.0   
4          18.0              Rex (voice)  52fe4284c3a36847f8024fa5     2.0   
...         ...                      ...                       ...     ...   
564887      3.0                           52fe4ea59251416c7515d7d9     1.0   
564888      4.0                           52fe4ea59251416c7515d7dd     2.0   
564889      5.0                           52fe4ea59251416c7515d7e1     0.0   
564890      6.0                           52fe4ea59251416c7515d7e5     1.0   
564891      NaN                      NaN                       NaN     NaN   

               id               name  order                      profile_path  
0            31.0          Tom Hanks    0.0  /pQFoyx7rp09CJTAb932F2g8Nlho.jpg  
1         12898.0          Tim Allen    1.0  /uX2xVf6pMmPepxnvFWyBtjexzgY.jpg  
2          7167.0        Don Rickles    2.0  /h5BcaDMPRVLHLDzbQavec4xfSdt.jpg  
3         12899.0         Jim Varney    3.0  /eIo2jVVXYgjDtaHoF19Ll9vtW7h.jpg  
4         12900.0      Wallace Shawn    4.0  /oGE6JqPP2xH4tNORKNqxbNPYi7u.jpg  
...           ...                ...    ...                               ...  
564887  1090923.0  Nathalie Lissenko    1.0                              None  
564888  1136422.0       Pavel Pavlov    2.0                              None  
564889  1261758.0  Aleksandr Chabrov    3.0                              None  
564890    29199.0        Vera Orlova    4.0  /n1NXVGNzNxtqsMWxLT1h8GO8Kpi.jpg  
564891        NaN                NaN    NaN                               NaN  

[564892 rows x 8 columns]

In [163]:
#Cambiamos nulos por listas vacias y evaluamos como diccionarios los datos de "crew"
data_credits['crew'] = data_credits['crew'].fillna('[not found]').apply(eval)

#Ahora crearemos un nuevo df para desanidar la columna "crew" 
df_crew = pd.json_normalize(data_credits['crew'].explode())

In [164]:
#Visualizamos el df del equipo de produccion
df_crew

credit_id  department  gender         id  \
0       52fe4284c3a36847f8024f49   Directing     2.0     7879.0   
1       52fe4284c3a36847f8024f4f     Writing     2.0    12891.0   
2       52fe4284c3a36847f8024f55     Writing     2.0        7.0   
3       52fe4284c3a36847f8024f5b     Writing     2.0    12892.0   
4       52fe4284c3a36847f8024f61     Writing     0.0    12893.0   
...                          ...         ...     ...        ...   
465080  52fe4776c3a368484e0c8399       Sound     0.0   549356.0   
465081  52fe4776c3a368484e0c839f      Camera     2.0    58818.0   
465082  533bccebc3a36844cf0011a7   Directing     0.0  1085341.0   
465083  58ebbc26925141281908aa0a  Production     2.0  1195656.0   
465084  593e676c92514105b702e68e   Directing     0.0  1344495.0   

                            job                 name  \
0                      Director        John Lasseter   
1                    Screenplay          Joss Whedon   
2                    Screenplay       Andrew Stanton   
3                    Screenplay           Joel Cohen   
4                    Screenplay         Alec Sokolow   
...                         ...                  ...   
465080  Original Music Composer       Richard McHugh   
465081  Director of Photography       João Fernandes   
465082                 Director     Yakov Protazanov   
465083                 Producer  Joseph N. Ermolieff   
465084                 Director        Daisy Asquith   

                            profile_path  
0       /7EdqiNbr4FRjIhKHyPPdFfEEEFG.jpg  
1       /dTiVsuaTVTeGmvkhcyJvKp2A5kr.jpg  
2       /pvQWsu0qc8JFQhMVJkTHuexUAa1.jpg  
3       /dAubAiZcvKFbboWlj7oXOkZnTSu.jpg  
4       /v79vlRYi94BZUQnkkyznbGUZLjT.jpg  
...                                  ...  
465080                              None  
465081                              None  
465082  /yyjbGdCs2ZN6IlZNCfmBWyuRDlt.jpg  
465083                              None  
465084                              None  

[465085 rows x 7 columns]

In [165]:
'''Se puede notar que hay datos que no serviran para cumplir las consignas del proyecto, por lo tanto solo vamos a conservar el nombre de los
miembros del reparto de actores que nos ayudará con uno de los endpoints del MVP'''
#Aplicamos la formula para extraer nombres y lo guardamos en la nueva columna "cast_name"
data_credits['cast'] = data_credits['cast'].apply(extraer_nombres)

#A los valores vacios que hayan quedado los vamos a volver nulos para luego en el EDA rellenarlos con "not found"
data_credits['cast'] = data_credits['cast'].replace([], np.nan)

In [166]:
'''Para el caso de la columna "crew", de todos los miembros del equipo de producción de la pelicula, solo nos sirve el dato del nombre del
director de la misma, entonces los vamos a conservar en una nueva columna llamada "director" y desestimaremos el resto de la informacion'''
#Creamos una función que nos ayude a extraer los nombres de los directores en "crew"
def extraer_director(crew):
    for miembro in crew:
        if miembro.get('job') == 'Director':
            return miembro.get('name')
    return None
#Aplicamos la función y lo guardamos en la nueva columna "Director"
data_credits['director'] = data_credits['crew'].apply(extraer_director)

#A los valores vacios que hayan quedado los vamos a volver nulos para luego en el EDA rellenarlos con "not found"
data_credits['director'] = data_credits['director'].replace([], np.nan)


In [167]:
#Eliminamos la columna original "crew"
data_credits = data_credits.drop(columns=['crew'])

In [168]:
#Visualizamos el df data_credits y su composicion hasta ahora con los cambios realizados
data_credits.head()

cast     id         director
0  [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...    862    John Lasseter
1  [Robin Williams, Jonathan Hyde, Kirsten Dunst,...   8844     Joe Johnston
2  [Walter Matthau, Jack Lemmon, Ann-Margret, Sop...  15602    Howard Deutch
3  [Whitney Houston, Angela Bassett, Loretta Devi...  31357  Forest Whitaker
4  [Steve Martin, Diane Keaton, Martin Short, Kim...  11862    Charles Shyer

In [169]:
'''Ahora vamos a unir el df credits al df movies con el metodo merge y usando el "id" de las 
peliculas, en modo "left" para que se unan solo el reparto y los directores de pelicula que esten
listadas en movies'''
#Aplicamos merge para la unión
movies_credits_final = pd.merge(data_movies, data_credits, on='id', how='left')

   EDA (Analisis exploratorio de datos)

Seguiremos con un analisis de la información que nos va quedando en el dataset unido aplicando algunas descripciones estadisticas a las caracteristicas numericas y a las categoricas, también veremos nuevamente los valores nulos, y buscaremos valores duplicados y veremos si aparecen valores atípicos

In [170]:
#Buscaremos un resumen estadistico de las variables numericas
movies_credits_final.describe()

budget             id    popularity  \
count  4.545200e+04   45452.000000  45452.000000   
mean   4.231201e+06  107991.064926      2.926791   
min    0.000000e+00       2.000000      0.000000   
25%    0.000000e+00   26367.750000      0.390000   
50%    0.000000e+00   59872.000000      1.130000   
75%    0.000000e+00  156330.500000      3.690000   
max    3.800000e+08  469172.000000    547.490000   
std    1.742923e+07  112118.576134      6.006419   

                        release_date       revenue       runtime  \
count                          45452  4.545200e+04  45206.000000   
mean   1992-05-16 05:08:17.720672384  1.121905e+07     94.179821   
min              1874-12-09 00:00:00  0.000000e+00      0.000000   
25%              1978-10-08 00:00:00  0.000000e+00     85.000000   
50%              2001-09-01 12:00:00  0.000000e+00     95.000000   
75%              2010-12-16 06:00:00  0.000000e+00    107.000000   
max              2020-12-16 00:00:00  2.787965e+09   1256.000000   
std                              NaN  6.433916e+07     38.331639   

       vote_average    vote_count        return  release_year  
count  45452.000000  45452.000000  4.545200e+04  45452.000000  
mean       5.623962    109.987283  6.589397e+02   1991.882579  
min        0.000000      0.000000  0.000000e+00   1874.000000  
25%        5.000000      3.000000  0.000000e+00   1978.000000  
50%        6.000000     10.000000  0.000000e+00   2001.000000  
75%        6.800000     34.000000  0.000000e+00   2010.000000  
max       10.000000  14075.000000  1.239638e+07   2020.000000  
std        1.915585    491.347202  7.463082e+04     24.057750

In [171]:
#Buscaremos un resumen estadistico de las variables categoricas
movies_credits_final.describe(include=['O'])

genres original_language            overview production_companies  \
count     45452             45441               44511                45452   
unique     4065                89               44232                22668   
top     [Drama]                en  No overview found.                   []   
freq       5006             32249                 133                11818   

              production_countries spoken_languages    status     title  \
count                        45452            45452     45372     45452   
unique                        2389             1843         6     42196   
top     [United States of America]        [English]  Released  Blackout   
freq                         17868            22410     45010        13   

         cast   director  
count   45451      44616  
unique  42657      17549  
top        []  John Ford  
freq     2354         66

In [172]:
#Evaluamos si hay 'id' duplicados, es decir, si hay peliculas duplicadas en el df final
movies_credits_final.duplicated(['id']).any()

np.True_

In [173]:
#Evaluamos los valores nulos/faltantes y su cantidad en el dataset
movies_credits_final.isnull().sum()

budget                    0
genres                    0
id                        0
original_language        11
overview                941
popularity                0
production_companies      0
production_countries      0
release_date              0
revenue                   0
runtime                 246
spoken_languages          0
status                   80
title                     0
vote_average              0
vote_count                0
return                    0
release_year              0
cast                      1
director                836
dtype: int64

CONCLUSIONES PRELIMINARES

* Podemos observar que en la columna "status" se repite en casi el 100% del dataset el valor "Released", por lo tanto podemos eliminarla mas aún teniendo en cuenta que todas las películas tienen su fecha de estreno.
* En las columna "original_languages" vemos que se repite el idioma inglés en un alto porcentaje, aún asi la consideramos candidata a ser desestimada porque no aporta informacion relevante para recomendar películas.
* Respecto de la columna "overview" vemos que contiene pocos nulos, los vamos a rellenar con "not found", y a los valores "No overview found" los reemplazaremos por "not found" también para que haya uniformidad.
* Vemos que en la columna "runtime" hay valores atipicos, no existe películas de 1256 minutos. También hay valores nulos, estos vamos a rellenarlos con la media de los valores existentes en la columna.
* Las columnas "return", "budget" y "revenue" tienen en su gran mayoria valores en 0, lo que nos da una idea que no serán caracteristicas que nos agreguen una buena prediccion o ajuste a nuestro modelo, en efecto, podrían ser desestimadas para el endpoint de ML.
* En las columnas de "cast" y "director" vemos que han quedado nulos luego de la union de los 2 datasets, los vamos a rellenar con "not found" para cumplir con la uniformidad de valores faltantes.
* Por último, podemos ver que hay duplicados de peliculas, por lo tanto las eliminaremos.


In [174]:
#Eliminamos "status" y "original_languages"
movies_credits_final = movies_credits_final.drop(columns=['status', 'original_language'])

#Completamos los valores nulos de "runtime" con la media
movies_credits_final['runtime'] = movies_credits_final['runtime'].fillna(movies_credits_final['runtime'].mean())

#Columna "overview" reemplazamos valores nulos y "No overview found" con "not found"
movies_credits_final['overview'] = movies_credits_final['overview'].replace('No overview found', 'not found')
movies_credits_final['overview'] = movies_credits_final['overview'].fillna('not found')

#Rellenamos valores nulos en "cast" y en "director"
movies_credits_final['cast'] = movies_credits_final['cast'].fillna('not found')
movies_credits_final['director'] = movies_credits_final['director'].fillna('not found')

'''Eliminamos las peliculas duplicadas en base a la columna 'id', y queda el df en condiciones para continuar con el EDA a más 
profundidad y poder cumplir las consignas del MVP'''
movies_credits_final = movies_credits_final.drop_duplicates(subset='id')

In [175]:
#Visualizamos como ha resultado el ETL hasta el momento
movies_credits_final.info()

<class 'pandas.core.frame.DataFrame'>
Index: 45346 entries, 0 to 45451
Data columns (total 18 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   budget                45346 non-null  float64       
 1   genres                45346 non-null  object        
 2   id                    45346 non-null  int64         
 3   overview              45346 non-null  object        
 4   popularity            45346 non-null  float64       
 5   production_companies  45346 non-null  object        
 6   production_countries  45346 non-null  object        
 7   release_date          45346 non-null  datetime64[ns]
 8   revenue               45346 non-null  float64       
 9   runtime               45346 non-null  float64       
 10  spoken_languages      45346 non-null  object        
 11  title                 45346 non-null  object        
 12  vote_average          45346 non-null  float64       
 13  vote_count           

In [176]:
#Exportamos el dataframe final a un csv
movies_credits_final.to_csv('dataset_final.csv')

In [177]:
'''Primer endpoint a desarrollar en la API:
def cantidad_filmaciones_mes( Mes ): Se ingresa un mes en idioma Español. Debe devolver la cantidad de
películas que fueron estrenadas en el mes consultado en la totalidad del dataset.'''

def cantidad_filmaciones_mes( Mes ):

    #Convertimos el nombre del mes en español a número
    meses = {
        'enero': 1,
        'febrero': 2,
        'marzo': 3,
        'abril': 4,
        'mayo': 5,
        'junio': 6,
        'julio': 7,
        'agosto': 8,
        'septiembre': 9,
        'octubre': 10,
        'noviembre': 11,
        'diciembre': 12
    }
    mes_a_numero = meses.get(Mes.lower())
    
    #Si no se ha ingresado un mes valido en español retornaremos el mensaje de advertencia
    if mes_a_numero not in [1,2,3,4,5,6,7,8,9,10,11,12] :
        return f"El mes '{Mes}' no es válido. Por favor ingresa un mes en español."
    
    #Si se ha ingresado un mes valido en español retornaremos la cantidad de filmaciones aplicando el siguiente filtro
    cantidad_pelis = movies_credits_final[movies_credits_final['release_date'].dt.month == mes_a_numero].shape[0]
    
    return f"En {Mes} fueron estrenadas {cantidad_pelis} películas."

In [178]:
#Hacemos una prueba a la funcion cantidad_filmaciones_mes
consulta_mes = cantidad_filmaciones_mes("mayo")
print(consulta_mes)

En mayo fueron estrenadas 3337 películas.


In [179]:
'''En el segundo endpoint se solicita:
def cantidad_filmaciones_dia( Dia ): Se ingresa un día en idioma Español. Debe devolver la cantidad de películas
que fueron estrenadas en día consultado en la totalidad del dataset.'''

def cantidad_filmaciones_dia(Dia):
    
    #Si el dia se debe ingresar en español tendremos que usar el metodo dayofweek en la consulta, para eso debemos mapear
    #los dias de la semana en su nombre en español a sus índices
    semanaDia = {"lunes": 0,"martes": 1,"miércoles": 2,"jueves": 3,"viernes": 4,"sábado": 5,"domingo": 6}

    #Pasamos a minuscula el dia ingresado en español en la consulta
    dia_minuscula = Dia.lower()

    #Si no se ha ingresado un día valido en español retornaremos el mensaje de advertencia
    if dia_minuscula not in semanaDia:
        return f"El día '{Dia}' no es válido. Por favor ingresar un día de la semana en español."
    
    #Nos quedamos con el indice del dia correspondiente
    indiceDia = semanaDia[dia_minuscula]

    #Ahora filtramos la cantidad de peliculas que se estrenaron en el dia ingresado
    cantidad_pelis = movies_credits_final[movies_credits_final['release_date'].dt.dayofweek == indiceDia].shape[0]

    return f"En los días {Dia} fueron estrenadas {cantidad_pelis} peliculas"

In [180]:
#Hacemos una prueba a la funcion cantidad_filmaciones_dia
consulta_dia = cantidad_filmaciones_dia("jueves")
print(consulta_dia)

En los días jueves fueron estrenadas 7520 peliculas


In [181]:
'''En el tercer endpoint se solicita:
def score_titulo( titulo_de_la_filmación ): Se ingresa el título de una filmación esperando como respuesta
el título, el año de estreno y el score.'''
def score_titulo(titulo_de_la_filmacion ):

   #Guardaremos en una variable los datos del df correpondiente al titulo ingresado. Aplicaremos minuscula
   pelicula = movies_credits_final[movies_credits_final['title'].str.lower() == titulo_de_la_filmacion.lower()]
   
   #Si no se ha encontrado la pelicula, devolvemos un mensaje de advertencia
   if pelicula.empty:
      return f"El nombre de la película '{titulo_de_la_filmacion}' no se encontró en la consulta. Por favor intente otro nombre."
    
   #Si la pelicula se ha encontrado, retornamos los datos solicitados
   año_estreno = pd.to_datetime(pelicula['release_date'].values[0]).year
   titulo = pelicula['title'].values[0]
   score = pelicula['popularity'].values[0]

   return f"La película '{titulo}' se ha estrenado el año {año_estreno} y cuenta con {score} de puntaje de popularidad asignado por TMDB"

In [182]:
#Hacemos una prueba a la funcion score_titulo
ver_peli = score_titulo('jumanji')
print(ver_peli)

La película 'Jumanji' se ha estrenado el año 1995 y cuenta con 17.02 de puntaje de popularidad asignado por TMDB


In [183]:
'''Para el cuarto endpoint se solicita:
def votos_titulo( titulo_de_la_filmación ): Se ingresa el título de una filmación esperando como respuesta el título, la cantidad
de votos y el valor promedio de las votaciones. La misma variable deberá de contar con al menos 2000 valoraciones, caso contrario,
debemos contar con un mensaje avisando que no cumple esta condición y que por ende, no se devuelve ningun valor.'''
def votos_titulo(titulo_de_la_filmacion):

   #Guardaremos en una variable los datos del df correpondiente al titulo ingresado. Aplicaremos minuscula
   pelicula = movies_credits_final[movies_credits_final['title'].str.lower() == titulo_de_la_filmacion.lower()]

   #Si no se ha encontrado la pelicula, devolvemos un mensaje de advertencia
   if pelicula.empty:
      return f"El nombre de la película '{titulo_de_la_filmacion}' no se encontró en la consulta. Por favor intente otro nombre."
   
   #Si la pelicula encontrada no tiene 2000 valoraciones o mas, informamos eso con un mensaje y pedimos otro titulo.
   if pelicula['vote_count'].values[0] < 2000:
      return f"La película '{titulo_de_la_filmacion}' no tiene al menos 2000 votos. Intente con otra pelicula por favor."
   
   cant_votos = pelicula['vote_count'].values[0]
   promedio_votos = pelicula['vote_average'].values[0]
   titulo = pelicula['title'].values[0]

   return f"La pelicula {titulo} tiene {promedio_votos} en promedio de votos y cuenta con un total de {cant_votos} valoraciones"

In [184]:
#Hacemos una prueba a la funcion votos_titulo
ver_pelicula = votos_titulo('men of honor')
print(ver_pelicula)

La película 'men of honor' no tiene al menos 2000 votos. Intente con otra pelicula por favor.


In [189]:
'''Para el quinto endpoint se solicita:
def get_actor( nombre_actor ): Se ingresa el nombre de un actor que se encuentre dentro de un dataset debiendo devolver el éxito del
mismo medido a través del retorno. Además, la cantidad de películas que en las que ha participado y el promedio de retorno.
La definición no deberá considerar directores.'''
def get_actor(nombre_actor):
   
   nombre_actor = [nombre_actor]

   #Guardamos en una variable las filas o las peliculas del df donde se encuentre al actor
   peliculas_actor = movies_credits_final[movies_credits_final['cast'].apply(lambda x: any(d in x for d in nombre_actor))]

   #Si el actor ingresado no esta en la consulta devolvemos un mensaje de advertencia
   if peliculas_actor.empty:
       return f"El actor {nombre_actor} no se ha encontrado. Por favor intente con otro nombre o intente escribiendo las primeras letras con mayuscula."
   
   #Creamos las variables que filtran la consulta para luego mostrar en el resultado de la funcion
   retorno_actor = round(peliculas_actor['return'].sum(),2)
   retorno_promedio = round(peliculas_actor['return'].mean(),2)
   pelis_cantidad =  peliculas_actor['id'].shape[0]
   nombre_actor = nombre_actor[0]

   return f"El actor {nombre_actor} participó en {pelis_cantidad} peliculas con un retorno total de {retorno_actor} en millones de dolares, y ha tenido un promedio de {retorno_promedio} en millones de dolares de retorno por pelicula."

In [190]:
#Hacemos una prueba a la funcion nombre_actor
ver_actor = get_actor('Tom Cruise')
print(ver_actor)

El actor Tom Cruise participó en 47 peliculas con un retorno total de 183.02 en millones de dolares, y ha tenido un promedio de 3.89 en millones de dolares de retorno por pelicula.


In [191]:
'''Para el sexto endpoint se solicita:
def get_director( nombre_director ): Se ingresa el nombre de un director que se encuentre dentro de un dataset debiendo devolver el
éxito del mismo medido a través del retorno. Además, deberá devolver el nombre de cada película con la fecha de lanzamiento, retorno
individual, costo y ganancia de la misma.'''
def get_director(nombre_director):

    #Guardamos en una variable las filas o las peliculas del df donde se encuentre al director
    peliculas_director = movies_credits_final[movies_credits_final['director'] == nombre_director]

    #Si el director ingresado no esta en la consulta devolvemos un mensaje de advertencia
    if peliculas_director.empty:
        return f"El director {nombre_director} no se ha encontrado. Por favor intente con otro nombre o intente escribiendo las primeras letras con mayuscula."
    
    #Creamos las variables que filtran la consulta para luego mostrar en el resultado de la funcion
    retorno_director = round(peliculas_director['return'].sum(),2)
    exito_director = f"El director {nombre_director} en sus peliculas tuvo un retorno total de {retorno_director} en millones de dolares"

    return exito_director, peliculas_director[['title', 'release_date', 'budget', 'return']]

In [192]:
#Hacemos una prueba a la funcion get_director
ver_director = get_director('James Cameron')
print(ver_director)

('El director James Cameron en sus peliculas tuvo un retorno total de 54.24 en millones de dolares',                                 title release_date       budget  return
375                         True Lies   1994-07-14  115000000.0    3.29
582        Terminator 2: Judgment Day   1991-07-01  100000000.0    5.20
1093                        The Abyss   1989-08-09   70000000.0    1.29
1159                           Aliens   1986-07-18   18500000.0    9.91
1196                   The Terminator   1984-10-26    6400000.0   12.25
1641                          Titanic   1997-11-18  200000000.0    9.23
5646   Piranha Part Two: The Spawning   1981-01-01     145786.0    0.00
6157              Ghosts of the Abyss   2003-04-10   13000000.0    1.31
9664               Aliens of the Deep   2005-01-21          0.0    0.00
14574                          Avatar   2009-12-10  237000000.0   11.76
26591                        Avatar 2   2020-12-16          0.0    0.00
43979      T2 3-D: Battle Across Ti